internal/expr/bindtypes.go

```diff
@@ -75,63 +72,68 @@ func (pe *ParsedExpr) BindTypes(args ...any) (te *TypedExpr, err error) {
 		}
 	}
 
-	var sql bytes.Buffer
-
-	var inCount int
-	var outCount int
-
-	var outputs = make([]typeinfo.Member, 0)
-	var inputs = make([]typeinfo.Member, 0)
-
-	var typeMemberPresent = make(map[typeinfo.Member]bool)
+	// Bind types to each expression.
+	typedExprs := []typedExpression{}
+	outputMemberUsed := map[typeinfo.Member]bool{}
+	for _, expr := range *pe {
+		te, err := expr.bindTypes(ti)
+		if err != nil {
+			return nil, err
+		}
 
-	// Check and expand each query expr.
-	for _, expr := range pe.exprs {
-		switch e := expr.(type) {
-		case *inputExpr:
-			inLoc, err := bindInputTypes(ti, e)
+		if toe, ok := te.(*typedOutputExpr); ok {
+			err = toe.checkUsed(outputMemberUsed)
 			if err != nil {
 				return nil, err
 			}
-			sql.WriteString("@sqlair_" + strconv.Itoa(inCount))
-			inCount++
-			inputs = append(inputs, inLoc)
-		case *outputExpr:
-			outCols, typeMembers, err := bindOutputTypes(ti, e)
-			if err != nil {
-				return nil, err
-			}
-
-			for _, tm := range typeMembers {
-				if ok := typeMemberPresent[tm]; ok {
-					return nil, fmt.Errorf("member %q of type %q appears more than once in output expressions", tm.MemberName(), tm.OuterType().Name())
-				}
-				typeMemberPresent[tm] = true
-			}
-
-			for i, c := range outCols {
-				sql.WriteString(c.String())
-				sql.WriteString(" AS ")
-				sql.WriteString(markerName(outCount))
-				if i != len(outCols)-1 {
-					sql.WriteString(", ")
-				}
-				outCount++
-			}
-			outputs = append(outputs, typeMembers...)
-		case *bypass:
-			sql.WriteString(e.chunk)
-		default:
-			return nil, fmt.Errorf("internal error: unknown query expr type %T", expr)
 		}
+
+		typedExprs = append(typedExprs, te)
 	}
 
-	return &TypedExpr{inputs: inputs, outputs: outputs, sql: sql.String()}, nil
+	typedExpr := TypedExprs(typedExprs)
+	return &typedExpr, nil
```

Comment:
Combine this into one line.

Reply:
Unfortunately this can't be, Go won't let you directly take the address when you do a type assertion